agent/command/macsign.go

```diff
@@ -0,0 +1,232 @@
+	if macSign.KeyId == "" {
+		catcher.New("key_id cannot be blank")
+	}
+	if macSign.Secret == "" {
+		catcher.New("secret cannot be blank")
+	}
+	if macSign.LocalZipFile == "" {
+		catcher.New("local_zip_file cannot be blank")
+	}
+	if macSign.OutputZipFile == "" {
+		catcher.New("output_zip_file cannot be blank")
+	}
+	if macSign.ServiceUrl == "" {
+		catcher.New("service_url cannot be blank")
+	}
+	if runtime.GOOS != "darwin" {
+		// do not fail, just set verifying to false.
+		macSign.Verify = false
+	}
+	if !(macSign.ArtifactType == "" || macSign.ArtifactType == "binary" || macSign.ArtifactType == "app") {
+		catcher.New(`artifact_type needs to be either blank, "binary" or "app"`)
+	}
+	if macSign.Notarize && macSign.BundleId == "" {
+		catcher.New("if notarization is requested, bundle_id cannot be blank")
+	}
+
+	return catcher.Resolve()
+}
+
+// Apply the expansions from the relevant task config
+// to all appropriate fields of the macSign.
+func (macSign *macSign) expandParams(conf *internal.TaskConfig) error {
+	var err error
+	if macSign.WorkingDir == "" {
+		macSign.WorkingDir = conf.WorkDir
+	}
+
+	if err = util.ExpandValues(macSign, conf.Expansions); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if !filepath.IsAbs(macSign.LocalZipFile) {
+		macSign.LocalZipFile = filepath.Join(macSign.WorkingDir, macSign.LocalZipFile)
+	}
+	if !filepath.IsAbs(macSign.OutputZipFile) {
+		macSign.OutputZipFile = filepath.Join(macSign.WorkingDir, macSign.OutputZipFile)
+	}
+	if macSign.EntitlementsFilePath != "" && !filepath.IsAbs(macSign.EntitlementsFilePath) {
+		macSign.EntitlementsFilePath = filepath.Join(macSign.WorkingDir, macSign.EntitlementsFilePath)
+	}
+	if macSign.ClientBinary != "" && !filepath.IsAbs(macSign.ClientBinary) {
+		macSign.ClientBinary = filepath.Join(macSign.WorkingDir, macSign.ClientBinary)
+	}
+
+	return nil
+}
+
+func (macSign *macSign) shouldRunForVariant(buildVariantName string) bool {
+	//If no buildvariant filter, run for everything
+	if len(macSign.BuildVariants) == 0 {
+		return true
+	}
+
+	//Only run if the buildvariant specified appears in the list.
+	return utility.StringSliceContains(macSign.BuildVariants, buildVariantName)
+}
+
+// Implementation of Execute. Expands the parameters,
+// and then execute macOS signing and/or notarization.
+func (macSign *macSign) Execute(ctx context.Context,
+	comm client.Communicator, logger client.LoggerProducer, conf *internal.TaskConfig) error {
+
+	if err := macSign.expandParams(conf); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := createEnclosingDirectoryIfNeeded(macSign.WorkingDir); err != nil {
+		return errors.Wrap(err, "problem making working directory")
+	}
+
+	if !macSign.shouldRunForVariant(conf.BuildVariant.Name) {
+		logger.Task().Infof("Skipping macsign of local file %s for variant %s",
+			macSign.LocalZipFile, conf.BuildVariant.Name)
+		return nil
+	}
+
+	// client binary
+	if macSign.ClientBinary == "" {
```

Comment:
Does it make sense to include this defaulting in in ExpandParams with the other defaults?